### DISCUSSION TITLE
How to draw a relationship diagram between photon beam brightness and photon energy

### DISCUSSION THREAD

**Discussion by lxdmay** [svenreiche/Genesis-1.3-Version4#188] [Q&A]
I would like to ask how to calculate photon  beam brightness and how to correctly select and calculate  photon  beam brightness from the data output by Genesis. I am a bit confused and do not have a specific idea to draw a relationship diagram between radiation photon  beam brightness and photon energy using Python or matlab. Could you please provide a specific example to illustrate? Thank you very much.

**Comment by svenreiche** [comment]
The photon brightness is defined as

number of photons per 
- size in x
- divergence in x'
- size in y
- divergence in y'
- duration t
- and bandwidth

The first 4 parameter you can get from the output file directly
Field/xsize, Field/xdivergence, Field/ysize, Field/ydivergence

then you have to calculate the rms duration of the radiation power profile and the rms bandwidth of the spectrum.

And the number of photon is the pulse energy devided by the photon energies.

Hope this helps

Sven

**Reply by lxdmay** [reply]
> rms bandwidth of the spectrum

  thanks, I am a little confused now, for example, if I want to know the photon beam brightness at z=1m position, which corresponds to multiple slices, then my field/xdivergence or ydivergence or xsize or ysize, for example, field/xdivergence, is taken from the value of a certain slice in xdivergence, or is it the sum of xdivergence corresponding to all slices, and then the average value is taken. Is bandwidth relative to bandwidth? That is to say, the RMS photon energy divided by the average photon energy is 0.1% BW. So the unit of photon beam brightness is photons/s/mm ^ 2/mrad ^ 2/0.1% BW.
  But I am not sure how to choose the data in xdivergence for two-dimensional matrices, and whether the bandwidth is calculated by dividing the root mean square photon energy by the average photon energy in the spectrum?

**Reply by lxdmay** [reply]
> The photon brightness is defined as
> 
> number of photons per
> 
> * size in x
> * divergence in x'
> * size in y
> * divergence in y'
> * duration t
> * and bandwidth
> 
> The first 4 parameter you can get from the output file directly Field/xsize, Field/xdivergence, Field/ysize, Field/ydivergence
> 
> then you have to calculate the rms duration of the radiation power profile and the rms bandwidth of the spectrum.
> 
> And the number of photon is the pulse energy devided by the photon energies.
> 
> Hope this helps
> 
> Sven

[brightness script.zip](https://github.com/user-attachments/files/16398827/brightness.script.zip)
According to you, I wrote the script for brightness. than my script is attatched here, and I hope you have time to take a look at it, thank you very much.

**Comment by lxdmay** [comment]
hello, Sven, I installed Genesis-1.3-Version4-4.6.6, but when I want to calculate the brightness, I find that the output file does not have the Field/xdivergence module, and the Field/xdivergence is not canceled in the setup settings, why is there no such module in the output file? The following is the setup and the output file "Field".

&setup
rootname = undulator
lattice = undulator.lat
beamline =undulator
lambda0=1e-9
gamma0=6653
delz=0.038
shotnoise=1
seed=1222
&end
![image](https://github.com/user-attachments/assets/6f0364e2-fa2f-48fb-8811-611597a64d7e)

**Comment by svenreiche** [comment]
The output of divergence require to have the FFTW library installed and Genesis compiled with it. Also you need to enable the output in the setup namelist.

**Reply by lxdmay** [reply]
thanks, If the FFTW library is compiled, which option is enabled in the setup namelist to output the divergence calculation? Previously, divergence was able to appear in the output file without any special settings.

**Reply by ZeugAusHH** [reply]
Hello,

the relevant switch is `exclude_fft_output` in the `&setup` namelist. If true, computation of divergence data is disabled.

Best
Christoph

**Reply by lxdmay** [reply]
> Hello,
> 
> the relevant switch is `exclude_fft_output` in the `&setup` namelist. If true, computation of divergence data is disabled.
> 
> Best Christoph

Christoph, thanks. I guess it's probably that my fftw didn't compile it well, because I didn't set up exclude_fft_output = true in the &setup namelist. Now my biggest wonder is whether this brightness script is written correctly. And I am anxious that this issue can be resolved. my script is attached here. 
[brightness script1.zip](https://github.com/user-attachments/files/16427971/brightness.script1.zip)

**Reply by ZeugAusHH** [reply]
Hello,

could it be that FFTW3 was not compiled in? Then the divergence code would not be included in your genesis binary.

If it is identified during the CMake run, you should get a message similar to
```
-- FFTW found
-- Version: 3.3.8
```

On the other hand, if it is not found, you would get something like:
```
-- FFTW not found
-- Excluding FFTW specific source code
```

Best
Christoph

**Reply by lxdmay** [reply]
> Hello,
> 
> could it be that FFTW3 was not compiled in? Then the divergence code would not be included in your genesis binary.
> 
> If it is identified during the CMake run, you should get a message similar to
> 
> ```
> -- FFTW found
> -- Version: 3.3.8
> ```
> 
> On the other hand, if it is not found, you would get something like:
> 
> ```
> -- FFTW not found
> -- Excluding FFTW specific source code
> ```
> 
> Best Christoph

thanks.
